src/logql-query.ts

```diff
@@ -1,48 +1,250 @@
+
+  syntaxTree.iterate({
+    from: node.from,
+    to: node.to,
+    enter(selectorsNode) {
+      if (selectorsNode.name === 'Identifier') {
+        label = query.slice(selectorsNode.from, selectorsNode.to);
+
+        return false;
+      } else if (['Eq', 'Neq', 'Re', 'Nre'].includes(selectorsNode.name)) {
+        operator = query.slice(selectorsNode.from, selectorsNode.to);
+
+        return false;
+      } else if (selectorsNode.name === 'String') {
+        value = query.slice(selectorsNode.from, selectorsNode.to);
+
+        return false;
+      }
+    },
+    leave(selectorsNode) {
+      if (selectorsNode.name === 'Matcher') {
+        matchers.push({ label, operator, value });
+      }
+    },
+  });
+
+  return matchers;
+};
+
+const parsePipelineStages = (
+  syntaxTree: Tree,
+  node: SyntaxNodeRef,
+  query: string,
+): Array<PipelineStage> => {
+  const pipelineStages: Array<PipelineStage> = [];
+  let operator: string,
+    value: string,
+    labelsInFilter: Array<LabelMatcher> | undefined;
+
+  syntaxTree.iterate({
+    from: node.from,
+    to: node.to,
+    enter(selectorsNode) {
+      if (selectorsNode.name === 'Pipe') {
+        operator = query.slice(selectorsNode.from, selectorsNode.to);
+
+        return false;
+      } else if (
+        ['PipeExact', 'PipeMatch', 'Nre', 'Neq'].includes(selectorsNode.name)
+      ) {
+        operator = query.slice(selectorsNode.from, selectorsNode.to);
+
+        return false;
+      } else if (
+        [
+          'LabelParser',
+          'JsonExpressionParser',
+          'LabelFilter',
+          'LineFormatExpr',
+          'LabelFormatExpr',
+          'String',
+        ].includes(selectorsNode.name)
+      ) {
+        value = query.slice(selectorsNode.from, selectorsNode.to);
+
+        if (selectorsNode.name === 'LabelFilter') {
+          labelsInFilter = parseMatchers(syntaxTree, selectorsNode, query);
+        }
+
+        return false;
+      }
+    },
+    leave(selectorsNode) {
+      if (selectorsNode.name === 'PipelineStage') {
+        pipelineStages.push({ operator, value, labelsInFilter });
+        labelsInFilter = undefined;
+      }
+    },
+  });
+
+  return pipelineStages;
+};
 
 export class LogQLQuery {
-  selectors: Array<string> = [];
-  pipeline: Array<string> = [];
+  streamSelector: Array<LabelMatcher> = [];
+  pipeline: Array<PipelineStage> = [];
 
   constructor(query: string) {
-    const parts = query.match(/.*{([^}]*)}(.*)/);
-    this.selectors =
-      parts && parts[1].length > 0
-        ? parts[1].split(',').filter(notEmptyString).map(trim)
-        : [];
-    this.pipeline =
-      parts && parts[2].length > 0
-        ? parts[2].split('|').filter(notEmptyString).map(trim)
-        : [];
+    const syntaxTree = parser.parse(query);
+    let parsedMatchers: Array<LabelMatcher> = [];
+    let parsedPipeline: Array<PipelineStage> = [];
+
+    syntaxTree.iterate({
+      enter(node) {
+        if (node.name) {
```

Comment:
Nit: The `if (node.name)` check looks redundant since we have `node.name === ...` in both of the branches below.